srttools/monitor/monitor.py

```diff
@@ -84,44 +81,73 @@ def _parse_filename(self, event):
         self._observer._timers[infile].start()
 
 
+class ProcessManager:
+    def __init__(self, maxsize):
+        self._semaphore = threading.Semaphore(maxsize)
+        self._lock = threading.Lock()
+        self._processes = set()
+        self._terminated = threading.Event()
+
+    def put(self, process, timeout=None):
+        if not self._semaphore.acquire(timeout=timeout) or self._terminated.is_set():
+            raise queue.Full
+        with self._lock:
+            self._processes.add(process)
+
+    def remove(self, process):
+        with self._lock:
+            if not self._terminated.is_set() and process in self._processes:
+                self._processes.remove(process)
+                self._semaphore.release()
+
+    def terminate(self):
+        with self._lock:
+            self._terminated.set()
+            for process in set(self._processes):
+                if process.is_alive():
+                    process.terminate()
+                self._processes.remove(process)
+                self._semaphore.release()
```

Comment:
Migrated from a queue.Queue to this custom class. I was using the queue improperly by accessing to inner members, which was probably not thread safe.